soc/riscv/riscv-privilege/telink_w91/ipc/src/ipc_dispatcher.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2023 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/ipc/ipc_service.h>
+#include <zephyr/device.h>
+#include <zephyr/logging/log.h>
+
+#include "ipc_dispatcher.h"
+
+#define LOG_LEVEL LOG_LEVEL_INFO
+#define LOG_MODULE_NAME ipc_dispatcher
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#define IPC_BOUND_TIMEOUT_IN_MS 		K_MSEC(1000)
+
+static struct ipc_ept ept;
+static struct k_mutex ipc_mutex;
+static sys_slist_t ipc_elem_list;
+static K_SEM_DEFINE(edp_bound_sem, 0, 1);
+
+static void endpoint_bound(void *priv)
+{
+	k_sem_give(&edp_bound_sem);
+}
+
+static void endpoint_received(const void *data, size_t len, void *priv)
+{
+	bool is_elem_added = false;
+	ipc_dispatcher_elem_t *p_elem;
+	ipc_dispatcher_id_t *p_id = (ipc_dispatcher_id_t *)data;
+
+	k_mutex_lock(&ipc_mutex, K_FOREVER);
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&ipc_elem_list, p_elem, node) {
+		if ((p_elem->id == *p_id) && (p_elem->cb != NULL)) {
+			is_elem_added = true;
+			break;
+		}
+	}
+
+	k_mutex_unlock(&ipc_mutex);
+
+	if (is_elem_added) {
+		p_elem->cb(data, len);
```

Comment:
under mutex. otherwise it can be deleted

Reply:
Updated